vyper/venom/README.md

```diff
@@ -160,3 +160,250 @@ A number of passes that are planned to be implemented, or are implemented for im
+
+### IRFunction
+An `IRFunction` is composed of a name and multiple `IRBasicBlocks`, with one marked as the entry point to the function.
+
+### IRBasicBlock
+An `IRBasicBlock` contains a label and a sequence of `IRInstructions`.
+Each `IRBasicBlock` has a single entry point and exit point.
+The exit point must be one of the following terminator instructions:
+- `jmp` 
+- `djmp` 
+- `jnz` 
+- `ret` 
+- `return` 
+- `stop` 
+- `exit`
+
+Normalized basic blocks can not have multiple predecessors and successors. It has either one (or zero) predecessors and potentially multiple successors or vice versa.
+
+### IRInstruction
+An `IRInstruction` consists of an opcode, a list of operands, and an optional return value.
+An operand can be a label, a variable, or a literal.
+
+## Instructions
+
+### Special instructions
+
+- `invoke`
+  - Cause control flow to jump to a function denoted by the label.
+  - Return values are passed in the return buffer at the offset address.
+  - Practically only used for internal functions.
+  - Effectively translates to `JUMP` and therefore changes the program counter value.
+  - ```
+    invoke offset, label
+    ```
+- `alloca`
+  - Allocates memory of a given size at a given offset in memory.
+  - The output is the offset itself.
+  - Because the SSA form does not allow changing values of registers, handling mutable variables can be tricky. The `alloca` instruction is meant to simplify that.
+  - ```
+    out = alloca size, offset
+    ```
+- `palloca`
+  - Like the `alloca` instruction but only used for parameters of internal functions.
+  - ```
+    out = palloca size, offset
+    ```
+- `iload`
+  - Load value at immutable section of memory denoted by `offset` into `out` variable.
+  - The operand can be either a literal, which is a statically computed offset, or a variable.
+  - ```
+    out = iload offset
+    ```
+- `istore`
+  - The instruction represents a store into immutable section of memory.
+  - Like in `iload`, the offset operand can be a literal.
+  - ```
+    istore offset value
+    ```
+- phi
+  - label, variable, basic phi
+  - ```
+    %out = phi %61:2, label_a, %61, label %__main_entry
+    ```
+- `offset`
+  - Statically compute offset. Useful for `mstore`, `mload` and such.
+  - Basically `label` + `op`.
+  - ```
+    ret = offset label, op
+    ```
+- `param`
+  - The `param` instruction is used to represent function arguments passed by the stack.
+  - We assume the argument is on the stack and the `param` instruction is used to ensure we represent the argument by the `out` variable.
+  - ```
+    out = param
+    ```
+- `store`
+  - Store variable value or literal into `out` variable.
+  - ```
+    out = op
+    ```
+- dbname
+  - make and mark a data segment (one data segment in context - so maybe section it?) dunno
+- db
+  - db stores into the data segment some label? hmm
+- dloadbytes
+  - aparently the same `codecopy`-everything handled the same way. Maybe historical reasons?
```

Comment:
it's handled differently in the legacy IR pipeline, but in venom they are handled the same